Core/src/IfcGeometryConverter/FaceConverter.h

```diff
@@ -835,65 +844,131 @@ namespace OpenInfraPlatform {
 
 					// get reference to IfcFaceBound's (attribute 1)
 					const std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>>& ifcFaceBounds = advancedFace->Bounds;
-					// if one face bound is not of the allowed type: throw; else: continue and use already converted faceBoundLoops
+					// if one face bound is not of the allowed type: throw; else: continue
 					if (!std::all_of(ifcFaceBounds.begin(), ifcFaceBounds.end(), [](const auto& ifcFaceBound) {return
 						ifcFaceBound->Bound.isOfType<typename IfcEntityTypesT::IfcEdgeLoop>() || ifcFaceBound->Bound.isOfType<typename IfcEntityTypesT::IfcVertexLoop>(); }))
 					{
 						throw oip::InconsistentModellingException(advancedFace, "IfcAdvancedFace has a loop type as face boundary which is not allowed.");
 					}
 
+					// get reference to IfcSurface (attribute 2)
+					const EXPRESSReference<typename IfcEntityTypesT::IfcSurface>& faceSurface = advancedFace->FaceSurface;
+					// in case of IfcAdvancedFace, the FaceSurface-entity is restricted to 3 types (respectively 7 subtypes; according to ifc 4x1)
+					if (!(faceSurface.isOfType<typename IfcEntityTypesT::IfcElementarySurface>()
+						|| faceSurface.isOfType<typename IfcEntityTypesT::IfcSweptSurface>()
+						|| faceSurface.isOfType<typename IfcEntityTypesT::IfcBSplineSurface>()))
+					{
+						throw oip::InconsistentModellingException(advancedFace, "IfcAdvancedFace has a surface type as face surface which is not allowed.");
+					}
+
+					computeIfcFaceSurface(advancedFace, pos, polyhedron, polyhedronIndices);
+				}
+
+				/*! \brief Computes the surface and boundary loop of an \c IfcFaceSurface (or sub-type), and puts them into one polyhedron.
+				\param[in]		ifcFaceSurface		\c IfcFaceSurface entity to be converted.
+				\param[in]		pos					The relative location of the origin of the representation's coordinate system within the geometric context.
+				\param[in,out]	polyhedron			Carve polyhedron of the converted face.
+				\param[in,out]	polyhedronIndices	Contains polyhedron indices of vertices (x,y,z converted to string).
+				*/
+				void computeIfcFaceSurface(

```

Comment:
There is already a function `convertIfcFaceSurface` within this pull request, about in line 807.
The idea behind: `convertIfcFaceSurface` will be called from `convertIfcFace` or from somewhere else. Depending on the ifc-entity, that function calls `convertIfcAdvancedFace` or hands over straight to this function `computeIfcFaceSurface`. As well, `convertIfcAdvancedFace` will hand over to `computeIfcFaceSurface`. `computeIfcFaceSurface` performes the actual calculation for both cases, IfcFaceSurface and IfcAdvancedFace. In the visualisation of doxygen, it looks the following:
![computeFaceSurface_callStructure](https://user-images.githubusercontent.com/64251870/146552738-18e22b56-7636-4eac-afae-10a94f14bdd2.png)
If we would apply the distinction of public and private functions in the geometry converter classes, `convertIfcFaceSurface` and `convertIfcAdvancedFace` would be public, and `computeIfcFaceSurface` would be private.

Reply:
Oh, I see. Please, do use the public/private distinction. You can put the function at the end of the file as a private function - this is already used in some places.

Reply:
Done in commit 0e39d35. I moved it into the protected section, in case another function will inherit from `FaceConverterT`.
Further functions moved to protected with commits 332747a and d6dbb6d.